[Yusuf Bin Musa] iP

yusufaine opened the pull request:
# DukePro
> "Your mind is for having ideas, not holding them." – David Allen ([Source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. 

It's:
* 💻  text-based
* 🧑‍🎓  easy to learn
* 🚀  ~FAST~ _SUPER_ FAST to use

All you need to do is,
1. download the latest version [here](https://github.com/yusufaine/ip/releases),
2. ~run it in your terminal~ double-click it 😏 
3. add your tasks
4. let it manage your tasks for you 😉 

And the best part -- it's **FREE**!

## Features:
- [x] ✅  Managing tasks
- [x] ⏰  Managing deadlines
- [x] 📅  Events or reminders

### Feature pipeline
<details>
<summary> Click here if you're interested to see what's in the pipeline 😉  </summary>

- [ ] Show Tasks based on their types (todo/deadlines/reminder)
- [ ] Sort events and deadlines by their stated time to occur/due
- [ ] Mark/unmark/remove several entries at once
- [ ] Add command aliasing (lookup table)
</details>


---
If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

yusufaine committed:
feat: changed greeting message

yusufaine committed:
feat: added padding to greeting

yusufaine committed:
feat: added greet, botResponse, and exit

yusufaine committed:
updated ignore to exclude .class

yusufaine committed:
refactor: structure of code, removed global vars

yusufaine committed:
feat: added display list, changed echo function to add to list

yusufaine committed:
feat: added Task class, added un-/mark feature

yusufaine committed:
feat: added task subclasses, updated output

yusufaine committed:
chore: forgot to include subclasses

yusufaine committed:
test: added semi-automated test

yusufaine committed:
feat: included error handling

yusufaine committed:
feat: added delete/remove feature, included clear to empty out the list

yusufaine committed:
chore: changed Task to be abstract to prevent creation of itself

yusufaine committed:
chore: moved response logic into enums

yusufaine committed:
chore, refactor: forgot to include enums file; refactored valid index logic in enums

yusufaine committed:
fix: subtask delimiter not specified, task initialised with task type

yusufaine committed:
fix: index/parseInt checks in place

yusufaine committed:
chore: switch-case convention

yusufaine committed:
feat: add local saving functionality

yusufaine committed:
update ignore, testing branching

yusufaine committed:
feat: event creation uses proper DateTime and displayed accordingly

yusufaine committed:
Merge branch 'master' into branch-level-7

# Conflicts:
#	.gitignore

yusufaine committed:
Merge branch 'master' into branch-level-8

yusufaine committed:
Merge branch 'branch-level-7'

yusufaine committed:
Merge branch 'branch-level-8'

yusufaine committed:
fix: date format difference between read and write, add todos

yusufaine committed:
refactor: cleanup  and

yusufaine committed:
test: include backup dat file for testing

yusufaine committed:
refactor: change how list is populated, add todo.dat

yusufaine committed:
refactor: update CRUD for data.dat

yusufaine committed:
refactor: major refactor for more OOP

yusufaine committed:
Merge branch 'branch-A-MoreOOP'

yusufaine committed:
add new OOP files

yusufaine committed:
refactor: packaged classes accordingly

yusufaine committed:
fix: empty command should not throw error

yusufaine committed:
test: add junit test for Storage and Parser, update .gitignore

yusufaine committed:
chore: update .gitignore with build artifact, update bot name and welcome message

yusufaine committed:
change bot name

yusufaine committed:
feat: add find command

yusufaine committed:
chore: fix coding standard

yusufaine committed:
docs: added javadoc

yusufaine committed:
feat: add find command, modified list to use lambda

yusufaine committed:
merge: branch 'branch-A-JavaDoc' to master

yusufaine committed:
resolve conflict between javadocs and coding standard

yusufaine committed:
feat (minor): add list alias ls

yusufaine committed:
feat(minor): included a list alias (ls)

consider: lookup table to include aliasing of existing commands

yusufaine committed:
chore: enforce coding standard

yusufaine committed:
Merge branch 'branch-A-CodingStandard'

Previous merge was not recorded

yusufaine committed:
feat: add clear task with confirmation

modify find command to use streams

DavidTan0527 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,104 @@
+     * @return the `Integer` value of the `String`.
+     * @throws DukeException if the parameter could not be parsed to an `Integer`.
+     */
+    public static int parseInt(String strToIntValue) throws DukeException {
```
Why did you create a method instead of using the built-in Java method [`Integer.parseInt(String s)`](https://docs.oracle.com/javase/10/docs/api/java/lang/Integer.html#parseInt(java.lang.String))?

yusufaine replied:
Regarding this point my rationale was that I want the errors to be thrown as and where they happen when I was implementing the parts that needed `parseInt()` which was `DeleteCommand`, and `ToggleCommand`. 

In those commands I've also thrown in out-of-bounds errors, in those 2 files I've also done some "error-checking" as UI was available to me there and I was not keen on passing UI into the Parser (since parser should only really be processing strings)

DavidTan0527 replied:
Ahh, I see, it makes more sense now. Thanks for the explanation!

DavidTan0527 commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,46 @@
+            StringBuilder sb = new StringBuilder(ui.STR_PADDING + "Here are the tasks in your list: \n");
+            int[] taskCount = {1};
+
+            tasks.getTasks().forEach( task -> {
+                sb.append(String.format(ui.STR_PADDING + "  %d. " + task, taskCount[0]++));
+
+                if (taskCount[0] != tasks.size() + 1) {
+                    sb.append("\n");
+                }
+            });
```
Perhaps, we can reinforce the concept of abstraction by putting this as a method in `TaskList` class

yusufaine replied:
Oh, I like this idea! Not sure why I didn't think of moving it into `TaskList`, but it makes more sense that way -- thank you!

DavidTan0527 commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,47 @@
+        if (tasks.isEmpty()) {
+            ui.printWithDivider("Your list is empty!");
+        } else {
+            List<Task> filteredTask = tasks.getTasks().stream()
+                    .filter(task -> task.getTaskName().contains(findString))
+                    .collect(Collectors.toList());
```
It might be better to abstract this method within `TaskList`, which might also work better with the code below.

yusufaine replied:
Similar to `ListCommand`, I think this is really helpful 👍

DavidTan0527 commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,47 @@
+                sb.append(ui.STR_PADDING + "Here are the matching tasks in your list:");
+
+                int[] taskCount = {1};
+                filteredTask.forEach(task -> {
+                    sb.append(String.format("\n" + ui.STR_PADDING + "  %d. " + task, taskCount[0]++));
+                });
```
Perhaps, we can reinforce the concept of abstraction by putting this as a method in `TaskList` class

DavidTan0527 commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,47 @@
+                StringBuilder sb = new StringBuilder(ui.STR_PADDING + "You've searched for: \"" + findString + "\"\n");
+                sb.append(ui.STR_PADDING + "Here are the matching tasks in your list:");
+
+                int[] taskCount = {1};
```
`int` would suffice.
```suggestion
                int taskCount = 1;
```

yusufaine replied:
I believe this would throw an error as I used a lambda to check for the filtered tasks, if this was a (enhanced) for loop, it would work, but it's not possible to change local variables in lambda as they are `final or effectively final` -- see screenshot attached.

![image](https://user-images.githubusercontent.com/29030941/151935959-1a30bad8-b515-4ed0-81d9-f869896d6f75.png)

DavidTan0527 replied:
Perhaps a traditional `for` loop would solve this issue 🤔

yusufaine replied:
yep, I could've done an enhanced for-loop (or a normal one) and iterate over, but part of me really wanted to use a lambda here 😆  I'm not quite sure what the consensus is about handling it that way, but to play it safe I'll probably change it to a for-loop -- thanks for the review!

DavidTan0527 commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,47 @@
+
+                int[] taskCount = {1};
+                filteredTask.forEach(task -> {
+                    sb.append(String.format("\n" + ui.STR_PADDING + "  %d. " + task, taskCount[0]++));
```
```suggestion
                    sb.append(String.format("\n" + ui.STR_PADDING + "  %d. " + task, taskCount++));
```

DavidTan0527 reviewed (commented):
Mostly great code design, just some suggestions that might help simplify your code. Please take a look. 😄

yusufaine committed:
refactor(minor): updated javadocs and isExit() in all commands

yusufaine committed:
Copying gradle files into branch-A-Gradle

yusufaine committed:
feat: add gradle support, specify to use java11

yusufaine committed:
Merge branch 'branch-A-Gradle'

yusufaine committed:
style: added checkstyle and fixed errors

yusufaine committed:
Merge branch 'branch-A-CheckStyle'

yusufaine committed:
style: fix checkstyle issue and enforce rules

yusufaine committed:
style: fix checkstyle issue and enforce rules

yusufaine committed:
feat!: reworked some of the app logic for GUI

fix checkstyle issues and removed unnecessary methods.

yusufaine committed:
Merge branch 'branch-A-Level10'

yusufaine committed:
chore: remove unnecessary assets

yusufaine committed:
cleanup: remove CLI test files

Remove CLI test files as our app now uses GUI

yusufaine committed:
feat: add assertion to ensure r/w access

add assertion to Duke to ensure null commands are not executed
add assertion to Storage to ensure r/w access to data file

fix buggy default font by specifying and including font to resources
change user image

yusufaine committed:
style: refactor according to CodeQuality

Unable to reduce the chunk in Parser::parseCommand
Storage::write has an arrowhead due to the required try-catch blocks

yusufaine committed:
refactor: use streams for iterating TaskList

this would also improve code readability by reducing for-loop blocks to
smaller and easier to read statements

yusufaine committed:
Merge pull request #1 from yusufaine/branch-A-Assertions

feat: add assertion to ensure r/w access

yusufaine committed:
Merge pull request #2 from yusufaine/branch-A-CodeQuality

style: refactor according to CodeQuality

yusufaine committed:
Merge branch 'branch-A-Streams'

yusufaine committed:
ci: create gradle.yml

yusufaine committed:
fix: add alias for delete to pass unit test

was initially removed to try and and keep the LOC of `parseCommand` < 30

yusufaine committed:
ci: fix gradle.yml

remove IO testing

yusufaine committed:
feat: add view function

include help function and update parserTest

yusufaine committed:
Merge pull request #4 from yusufaine/branch-B-ViewSchedules

feat: add view schedule, help, and better search

yusufaine committed:
feat: implement mass deletion

allows users to delete more than 1 task at a time so that they dont have
to key in the same command multiple times should they want to clear a
few tasks instead of nuking it with `clean`.

yusufaine committed:
Merge pull request #5 from yusufaine/branch-C-MassOps

feat: implement mass deletion

yusufaine committed:
remove misplaced file

yusufaine committed:
fix saving bug, will continue to monitor

yusufaine committed:
fix: shutdown bug not showing exit message

clean up the ExitCommand logic to give the users time to read the exit
message before saving the list and closing the application

yusufaine committed:
style: modify the UI to look nicer

yusufaine committed:
style: reduce space between messages

yusufaine committed:
docs: update README.md

yusufaine committed:
release: v0.1.3

yusufaine committed:
refactor: improve readability of DialogBox

yusufaine committed:
docs: update JavaDocs

yusufaine committed:
fix: update .gitignore and docs

Closes #6, and #7.

yusufaine committed:
Merge pull request #8 from yusufaine/branch-docs-01

fix: update .gitignore and docs

yusufaine committed:
chore: update .gitignore

yusufaine committed:
chore: test ignore

yusufaine committed:
chore: update .gitignore

yusufaine committed:
chore: update .gitignore
